### ISSUE TITLE
Issues with viewportRowRenderingOffset:auto while implementing custom expand collapse logic

### ISSUE THREAD

**Issue by samsonkhisty** [handsontable/handsontable#3192]
Hi there again, 

I'm trying to implement my expand collapse logic like so

![screen shot 2016-02-04 at 11 15 36 am](https://cloud.githubusercontent.com/assets/3254606/12821270/ab45793a-cb30-11e5-8220-3a62269447f1.png)

I've added a click event handler on the icons which will simply update a field `hidden_children` to the parent and `hidden` to all the children below and call `render` and I have a custom rendered that will change the icon on the parent and adds class `hide` to the children with `hidden:true`

I've got the bulk of the feature working, however ran into some issues: 

* If I leave the `viewportRowRenderingOffset: auto` as default, and collapse some rows in my view such that the rows which are normally outside the viewport are now into view. They are not rendered yet. 

* My next issue is again with  `viewportRowRenderingOffset: auto`, It introduces some weird scroll behavior on cell clicks if some data is collapsed. Not sure how to debug into this: 
![viewportrowrenderingoffset](https://cloud.githubusercontent.com/assets/3254606/12821909/7cf15646-cb33-11e5-9bed-55796fd75bba.gif)

* If I change to `viewportRowRenderingOffset: myData.length`, this works as intended however, for a table with huge rows there is lag in collapsing and expanding. It seems slow. 

Thanks in advance.

**Comment by AMBudnik** [comment]
Hi @samsonkhisty 
Great feature you're working on! I really would like to take a little part in your progress but without a demo it is very hard. Can you create one (for example in JSBin or JSFiddle)?

**Comment by samsonkhisty** [comment]
Hi @AMBudnik 
Finally I have a Demo for you. 

https://jsfiddle.net/fuggfuggfugg/nvd2q1nw/20/

In this example, I have row 5-9 hidden. Now if you click on row 24, cell 0. you will see that it performs the click and scrolldown so fast that it is multiselecting the rows. Here is a gif to replicate. 

![bug](https://cloud.githubusercontent.com/assets/3254606/12952222/beb194ba-cfe3-11e5-8e0b-65f8fee57f89.gif)


If you need more information. let me know.

**Comment by AMBudnik** [comment]
Thanks for sharing @samsonkhisty I did a little peek but it's hard to guess what's causing it. I'll mark this issue for other team members to look at.

**Comment by budnix** [comment]
Hi @samsonkhisty, Great job you have done! About the first screen it looks very cool. Issue you described is related with your technique you use to hide rows. It is not recommend to hide a lot of rows via css. In this case table calculates rows height even for this hidden rows. It reads height of hidden rows from cache and tries to fill whole viewport. The case is this row is hidden from DOM so that its why there is a gap.

For hiding rows you can use `modifyRow` hook and there put your logic. [In updated demo](https://jsfiddle.net/nvd2q1nw/24/) I hide rows indexes 2, 3 and rows from 10 to 14.

If you want to you can try to similar functionality which is [trimRows](http://docs.handsontable.com/pro/demo-trimming-rows.html). This plugin is part of Handsontable PRO.

**Comment by samsonkhisty** [comment]
@AMBudnik 

I still haven't been able to fix the issue using modifyRow, my CPU is through the roof. 
Is there a way I can try the PRO version trimRows and see if my code works, I can then proceed to purchase. 

Thanks.

**Comment by samsonkhisty** [comment]
@AMBudnik 

I still haven't been able to fix the issue using modifyRow, my CPU is through the roof. 
Is there a way I can try the PRO version trimRows and see if my code works, I can then proceed to purchase. 

Thanks.

**Comment by samsonkhisty** [comment]
@AMBudnik , @budnix

![bug](https://cloud.githubusercontent.com/assets/3254606/13586101/1a14f4d8-e48d-11e5-91ad-7b722aa9d9da.gif)

Finally managed to solve the issue after weeks of tinkering. we literally tried everything from hiding rows to the methods mentioned in the blog post. https://handsontable.com/blog/articles/3-ways-to-hide-rows-in-handsontable.

The modifyRow worked although it took a performance hit for some reason. My projects were big and the modifyRow was called a lot of times. The other issue with modifyRow was the rendering. I do indentation based on the sourceData for the row. During rendering I got the hidden data. Check jsFiddle in #3305 for more info.

My colleague went back to the first version where we had the scrolling issue, He finally was able to track and found that when clicking lower down the viewport, it thought that the cell was out of view and it tried to scroll. He then tracked it down to the cell height calculation bit and figured somehow if we could return a 0 for these hidden cells instead of the default height. He was lucky to find the modifyRowHeight would help calculate it correctly. 

I then hooked up the modifyRowHeight to return 1 when cells are collapse, I had to add additional CSS eventually get that effect in the gif. It gives a good indication as well of how many rows are collapsed. Overall I think I'm happy with this for now. 

This is an awesome library.
